File: src/Golang/disk/lib_test.go
package disk

import (
	"github.com/stretchr/testify/assert"
	"os"
	"testing"
)

func initDiskFileManager() *DiskFileManager {
	provider := SystemPathProvider{}
	dirManager := InitDiskDirectoryManager(&provider)

	fileManager := InitDiskFileManager(&provider, &dirManager)

	return &fileManager
}

func initDiskDirManager() *DiskDirectoryManager {
	provider := SystemPathProvider{}
	dirManager := InitDiskDirectoryManager(&provider)

	return &dirManager
}

func TestSystemPathProvider_CombineCombine(t *testing.T) {
	var res string
	var err error
	provider := new(SystemPathProvider)
	res, err = provider.Combine("parent", "dir", "file")
	assert.Nil(t, err)
	assert.Equal(t, "parent/dir/file", res)

	res, err = provider.Combine("parent/dir", "file")
	assert.Nil(t, err)
	assert.Equal(t, "parent/dir/file", res)

	res, err = provider.Combine("parent", "dir/file")
	assert.Nil(t, err)
	assert.Equal(t, "parent/dir/file", res)

	res, err = provider.Combine("parent/dir/file")
	assert.Nil(t, err)
	assert.Equal(t, "parent/dir/file", res)

	res, err = provider.Combine("/parent", "dir", "file")
	assert.Nil(t, err)
	assert.Equal(t, "/parent/dir/file", res)

	res, err = provider.Combine("/parent", "/dir", "file")
	assert.Nil(t, err)
	assert.Equal(t, "/parent/dir/file", res)

	res, err = provider.Combine("/parent/dir/file", "../..")
	assert.Nil(t, err)
	assert.Equal(t, "/parent", res)
}

func TestNormalizePath(t *testing.T) {
	var res string

	res = normalizePath("dir/file")
	assert.Equal(t, "dir/file", res)

	res = normalizePath("dir/file/")
	assert.Equal(t, "dir/file", res)

	res = normalizePath("dir/file//")
	assert.Equal(t, "dir/file", res)
}

func TestSystemPathProvider_GetObjectName(t *testing.T) {
	var res string
	var err error

	provider := new(SystemPathProvider)
	res, err = provider.GetObjectName("parent/dir/file")
	assert.Nil(t, err)
	assert.Equal(t, "file", res)

	res, err = provider.GetObjectName("parent/dir/")
	assert.Nil(t, err)
	assert.Equal(t, "dir", res)

	res, err = provider.GetObjectName("parent/dir//")
	assert.Nil(t, err)
	assert.Equal(t, "dir", res)

	res, err = provider.GetObjectName("parent/dir/../")
	assert.Nil(t, err)
	assert.Equal(t, "parent", res)

	res, err = provider.GetObjectName("/")
	assert.Equal(t, "", res)
	assert.NotNil(t, res)
	assert.Contains(t, err.Error(), "invalid path:")
}

func TestSystemPathProvider_GetObjectParentPath(t *testing.T) {
	var res string
	var err error

	provider := new(SystemPathProvider)
	res, err = provider.GetObjectParentPath("parent/dir/file")
	assert.Nil(t, err)
	assert.Equal(t, "parent/dir/", res)

	res, err = provider.GetObjectParentPath("parent/dir/")
	assert.Nil(t, err)
	assert.Equal(t, "parent/", res)

	res, err = provider.GetObjectParentPath("parent/dir//")
	assert.Nil(t, err)
	assert.Equal(t, "parent/", res)

	res, err = provider.GetObjectParentPath("/parent/dir/../")
	assert.Nil(t, err)
	assert.Equal(t, "/", res)

	res, err = provider.GetObjectParentPath("/")
	assert.Equal(t, "", res)
	assert.NotNil(t, res)
	assert.Contains(t, err.Error(), "invalid path:")

	res, err = provider.GetObjectParentPath("parent/dir/../")
	assert.Equal(t, "", res)
	assert.NotNil(t, res)
	assert.Contains(t, err.Error(), "invalid path:")
}

func TestDiskFileManager_GetPathProvider(t *testing.T) {
	provider := SystemPathProvider{}
	dirManager := InitDiskDirectoryManager(&provider)

	fileManager := InitDiskFileManager(&provider, &dirManager)

	assert.Equal(t, &provider, fileManager.GetPathProvider())
}

func TestDiskFileManager_GetDirectoryManager(t *testing.T) {
	provider := SystemPathProvider{}
	dirManager := InitDiskDirectoryManager(&provider)

	fileManager := InitDiskFileManager(&provider, &dirManager)

	assert.Equal(t, &dirManager, fileManager.GetDirectoryManager())
}

func TestDiskFileManager_CreateFile_WhenNotExists(t *testing.T) {
	fm := initDiskFileManager()

	// Create file...
	fd, err := fm.CreateFile("test_file")
	assert.Nil(t, err)
	assert.Equal(t, "test_file", fd.Name)
	assert.Equal(t, "", fd.Path)
	assert.Equal(t, int64(0), fd.Length)

	fullPath, err := fd.FullPath()
	assert.Nil(t, err)
	assert.Equal(t, "test_file", fullPath)

	// Cleanup
	_ = fm.DeleteFile("./test_file")
}

func TestDiskFileManager_CreateFile_WhenExists(t *testing.T) {
	fm := initDiskFileManager()

	// Preparing file...
	_, _ = fm.CreateFile("./test_file")

	// Recreate same file
	fd, err := fm.CreateFile("test_file")
	assert.Nil(t, fd)
	assert.NotNil(t, err)
	assert.True(t, os.IsExist(err))

	// Cleanup
	_ = fm.DeleteFile("./test_file")
}

func TestDiskFileManager_GetFile_WhenNotExists(t *testing.T) {
	fm := initDiskFileManager()

	fd, err := fm.GetFile("./test_file")
	assert.Nil(t, fd)
	assert.NotNil(t, err)
	assert.True(t, os.IsNotExist(err))
}

func TestDiskFileManager_GetFile_WhenExists(t *testing.T) {
	fm := initDiskFileManager()

	// Prepare file...
	_, _ = fm.CreateFile("./test_file")

	fd, err := fm.GetFile("test_file")
	assert.Nil(t, err)
	assert.Equal(t, "test_file", fd.Name)
	assert.Equal(t, "", fd.Path)
	assert.Equal(t, int64(0), fd.Length)

	fullPath, err := fd.FullPath()
	assert.Nil(t, err)
	assert.Equal(t, "test_file", fullPath)

	// Cleanup...
	_ = fm.DeleteFile("test_file")
}

func TestDiskFileManager_FileExists_WhenNotExists(t *testing.T) {
	fm := initDiskFileManager()

	exists, err := fm.FileExists("./test_file")
	assert.Nil(t, err)
	assert.False(t, exists)
}

func TestDiskFileManager_FileExists_WhenExists(t *testing.T) {
	fm := initDiskFileManager()

	// Prepare file...
	_, _ = fm.CreateFile("./test_file")

	exists, err := fm.FileExists("test_file")
	assert.Nil(t, err)
	assert.True(t, exists)

	// Cleanup...
	_ = fm.DeleteFile("./test_file")
}

func TestDiskFileManager_DeleteFile_WhenNotExists(t *testing.T) {
	fm := initDiskFileManager()

	err := fm.DeleteFile("./test_file")
	assert.NotNil(t, err)
	assert.True(t, os.IsNotExist(err))
}

func TestDiskFileManager_DeleteFile_WhenExists(t *testing.T) {
	fm := initDiskFileManager()

	// Prepare file...
	_, _ = fm.CreateFile("./test_file")

	err := fm.DeleteFile("test_file")
	assert.Nil(t, err)
}

func TestDiskDirectoryManager_GetPathProvider(t *testing.T) {
	provider := SystemPathProvider{}

	dm := InitDiskDirectoryManager(&provider)

	assert.Equal(t, &provider, dm.GetPathProvider())
}

func TestDiskDirectoryManager_CreateDir_WhenNotExists(t *testing.T) {
	dm := initDiskDirManager()

	dd, err := dm.CreateDir("test_dir")
	assert.Nil(t, err)
	assert.Equal(t, "test_dir", dd.Name)
	assert.Equal(t, "", dd.Path)

	fullPath, err := dd.FullPath()
	assert.Nil(t, err)
	assert.Equal(t, "test_dir", fullPath)

	// Cleanup...
	_ = dm.DeleteDir("./test_dir")
}

func TestDiskDirectoryManager_CreateDir_WhenExists(t *testing.T) {
	dm := initDiskDirManager()

	// Prepare dir...
	_, _ = dm.CreateDir("./test_dir")

	dd, err := dm.CreateDir("test_dir")
	assert.Nil(t, dd)
	assert.NotNil(t, err)
	assert.True(t, os.IsExist(err))

	// Cleanup...
	_ = dm.DeleteDir("./test_dir")
}

func TestDiskDirectoryManager_GetDir_WhenNotExists(t *testing.T) {
	dm := initDiskDirManager()

	dd, err := dm.GetDir("test_dir")
	assert.Nil(t, dd)
	assert.NotNil(t, err)
	assert.True(t, os.IsNotExist(err))
}

func TestDiskDirectoryManager_GetDir_WhenExists(t *testing.T) {
	dm := initDiskDirManager()

	// Prepare dir...
	_, _ = dm.CreateDir("./test_dir")

	dd, err := dm.GetDir("test_dir")
	assert.Nil(t, err)
	assert.Equal(t, "test_dir", dd.Name)
	assert.Equal(t, "", dd.Path)

	fullPath, err := dd.FullPath()
	assert.Nil(t, err)
	assert.Equal(t, "test_dir", fullPath)

	// Cleanup...
	_ = dm.DeleteDir("test_dir")
}

func TestDiskDirectoryManager_DirExists_WhenNonExists(t *testing.T) {
	dm := initDiskDirManager()

	exists, err := dm.DirExists("test_dir")
	assert.Nil(t, err)
	assert.False(t, exists)
}

func TestDiskDirectoryManager_DirExists_WhenExists(t *testing.T) {
	dm := initDiskDirManager()

	// Prepare dir...
	_, _ = dm.CreateDir("test_dir")

	exists, err := dm.DirExists("test_dir")
	assert.Nil(t, err)
	assert.True(t, exists)

	// Cleanup...
	_ = dm.DeleteDir("test_dir")
}

func TestDiskDirectoryManager_DeleteDir_WhenNotExists(t *testing.T) {
	dm := initDiskDirManager()

	err := dm.DeleteDir("test_dir")
	assert.NotNil(t, err)
	assert.True(t, os.IsNotExist(err))
}

func TestDiskDirectoryManager_DeleteDir_WhenExists(t *testing.T) {
	dm := initDiskDirManager()

	// Prepare dir...
	_, _ = dm.CreateDir("test_dir")

	err := dm.DeleteDir("test_dir")
	assert.Nil(t, err)
}

File: src/Rust/src/disk/lib_test.rs
use tempfile::{tempdir, TempDir};
use crate::providers::{DirectoryManager, FileManager, PathProvider};
use super::{SystemPathProvider, normalize_path, DiskDirectoryManager, DiskFileManager};

mod test_system_path_provider {
    use super::*;

    #[test]
    fn test_system_path_provider_combine() {
        let provider = SystemPathProvider::new();

        let res = provider.combine(vec!["parent", "dir", "file"]);
        assert!(res.is_ok());
        assert_eq!("parent/dir/file", res.unwrap());

        let res = provider.combine(vec!["parent/dir", "./file"]);
        assert!(res.is_ok());
        assert_eq!("parent/dir/file", res.unwrap());

        let res = provider.combine(vec!["parent", "dir/file"]);
        assert!(res.is_ok());
        assert_eq!("parent/dir/file", res.unwrap());

        let res = provider.combine(vec!["parent/dir/file"]);
        assert!(res.is_ok());
        assert_eq!("parent/dir/file", res.unwrap());

        let res = provider.combine(vec!["/parent", "dir", "file"]);
        assert!(res.is_ok());
        assert_eq!("/parent/dir/file", res.unwrap());

        let res = provider.combine(vec!["/parent/dir/file", "../.."]);
        assert!(res.is_ok());
        assert_eq!("/parent", res.unwrap());

        let res = provider.combine(vec!["parent", "dir", "../file", "../"]);
        assert!(res.is_ok());
        assert_eq!("parent/", res.unwrap());
    }

    #[test]
    fn test_system_path_provider_get_object_name() {
        let provider = SystemPathProvider::new();

        let res = provider.get_object_name("parent/dir/file");
        assert!(res.is_ok());
        assert_eq!("file", res.unwrap());

        let res = provider.get_object_name("parent/dir/");
        assert!(res.is_ok());
        assert_eq!("dir", res.unwrap());

        let res = provider.get_object_name("parent/dir//");
        assert!(res.is_ok());
        assert_eq!("dir", res.unwrap());

        let res = provider.get_object_name("parent/dir/../");
        assert!(res.is_ok());
        assert_eq!("parent", res.unwrap());

        let res = provider.get_object_name("/");
        assert!(res.is_err());
    }

    #[test]
    fn test_system_path_provider_get_object_parent_path() {
        let provider = SystemPathProvider::new();

        let res = provider.get_object_parent_path("parent/dir/file");
        assert!(res.is_ok());
        assert_eq!("parent/dir", res.unwrap());

        let res = provider.get_object_parent_path("parent/dir/");
        assert!(res.is_ok());
        assert_eq!("parent", res.unwrap());

        let res = provider.get_object_parent_path("parent/dir//");
        assert!(res.is_ok());
        assert_eq!("parent", res.unwrap());

        let res = provider.get_object_parent_path("/parent/dir/../");
        assert!(res.is_ok());
        assert_eq!("/", res.unwrap());

        let res = provider.get_object_parent_path("/");
        assert!(res.is_err());

        let res = provider.get_object_parent_path("parent/dir/../");
        assert!(res.is_err());
    }
}

#[test]
fn test_normalize_path() {
    assert_eq!("parent/dir", normalize_path("parent/dir/.".to_string()));

    assert_eq!("parent/dir/.file", normalize_path("parent/dir/.file".to_string()));

    assert_eq!("parent", normalize_path("parent/dir/..".to_string()));

    assert_eq!("parent", normalize_path("parent/dir/file/../..".to_string()));

    assert_eq!("parent/dir/..sub", normalize_path("parent/dir/file/../..sub".to_string()));

    assert_eq!("parent/dir/file", normalize_path("parent/dir/file/./.".to_string()));

    assert_eq!("parent/file", normalize_path("parent/dir/../file".to_string()));

    assert_eq!("parent/dir/file", normalize_path("parent/dir/./file".to_string()));

    assert_eq!("../parent/dir", normalize_path("../parent/dir".to_string()));

    assert_eq!("./parent/dir", normalize_path("./parent/dir".to_string()));

    assert_eq!(".parent/dir", normalize_path(".parent/dir".to_string()));

    assert_eq!("..parent/dir", normalize_path("..parent/dir".to_string()));

    assert_eq!("/parent/dir", normalize_path("/./parent/dir".to_string()));

    assert_eq!("/../parent/dir", normalize_path("/../parent/dir".to_string()));
}

mod test_disk_dir_manager_test {
    use super::*;

    #[test]
    fn test_disk_dir_manager_path_provider() {
        let spp = SystemPathProvider::new();

        let dm = DiskDirectoryManager::new(spp);

        let provider = dm.path_provider().as_any()
            .downcast_ref::<SystemPathProvider>()
            .unwrap();
        assert_eq!(&spp, provider);
    }

    #[tokio::test]
    async fn test_disk_dir_manager_create_dir_async_when_not_exists() {
        let dm = init_disk_dir_manager();

        let (tmp_dir, test_dir, _t) = get_test_dir(dm.path_provider());

        let res = dm.create_dir_async(&test_dir).await;
        assert!(res.is_ok());
        let dd = res.unwrap();
        assert_eq!("test_dir", dd.name);
        assert_eq!(tmp_dir, dd.path);
    }

    #[tokio::test]
    async fn test_disk_dir_manager_create_dir_async_when_exists() {
        let dm = init_disk_dir_manager();

        let (_, test_dir, _t) = get_test_dir(dm.path_provider());
        dm.create_dir_async(&test_dir).await.unwrap();

        let res = dm.create_dir_async(&test_dir).await;
        assert!(res.is_err());
    }

    #[tokio::test]
    async fn test_disk_dir_manager_get_dir_async_when_not_exists() {
        let dm = init_disk_dir_manager();

        let (_, test_dir, _t) = get_test_dir(dm.path_provider());

        let res = dm.get_dir_async(&test_dir).await;
        assert!(res.is_err());
    }

    #[tokio::test]
    async fn test_disk_dir_manager_get_dir_async_when_exists() {
        let dm = init_disk_dir_manager();

        let (tmp_dir, test_dir, _t) = get_test_dir(dm.path_provider());
        dm.create_dir_async(&test_dir).await.unwrap();

        let res = dm.get_dir_async(&test_dir).await;
        assert!(res.is_ok());
        let dd = res.unwrap();
        assert_eq!("test_dir", dd.name);
        assert_eq!(tmp_dir, dd.path);
    }

    #[tokio::test]
    async fn test_disk_dir_manager_is_dir_exists_async_when_not_exists() {
        let dm = init_disk_dir_manager();

        let (_, test_dir, _t) = get_test_dir(dm.path_provider());

        let res = dm.is_dir_exists_async(&test_dir).await;
        assert!(res.is_ok());
        assert!(!res.unwrap());
    }

    #[tokio::test]
    async fn test_disk_dir_manager_is_dir_exists_async_when_exists() {
        let dm = init_disk_dir_manager();

        let (_, test_dir, _t) = get_test_dir(dm.path_provider());
        dm.create_dir_async(&test_dir).await.unwrap();

        let res = dm.is_dir_exists_async(&test_dir).await;
        assert!(res.is_ok());
        assert!(res.unwrap());
    }

    #[tokio::test]
    async fn test_disk_dir_manager_delete_dir_async_when_not_exists() {
        let dm = init_disk_dir_manager();

        let (_, test_dir, _t) = get_test_dir(dm.path_provider());

        let res = dm.delete_dir_async(&test_dir, true).await;
        assert!(res.is_err());
    }

    #[tokio::test]
    async fn test_disk_dir_manager_delete_dir_async_when_exists() {
        let dm = init_disk_dir_manager();

        let (_, test_dir, _t) = get_test_dir(dm.path_provider());
        dm.create_dir_async(&test_dir).await.unwrap();

        let res = dm.delete_dir_async(&test_dir, true).await;
        assert!(res.is_ok());

        let (_, test_dir, _t1) = get_test_dir(dm.path_provider());
        dm.create_dir_async(&test_dir).await.unwrap();

        let res = dm.delete_dir_async(&test_dir, false).await;
        assert!(res.is_ok());
    }

    #[test]
    fn test_disk_dir_manager_create_dir_when_not_exists() {
        let dm = init_disk_dir_manager();

        let (tmp_dir, test_dir, _t) = get_test_dir(dm.path_provider());

        let res = dm.create_dir(&test_dir);
        assert!(res.is_ok());
        let dd = res.unwrap();
        assert_eq!("test_dir", dd.name);
        assert_eq!(tmp_dir, dd.path);
    }

    #[test]
    fn test_disk_dir_manager_create_dir_when_exists() {
        let dm = init_disk_dir_manager();

        let (_, test_dir, _t) = get_test_dir(dm.path_provider());
        dm.create_dir(&test_dir).unwrap();

        let res = dm.create_dir(&test_dir);
        assert!(res.is_err());
    }

    #[test]
    fn test_disk_dir_manager_get_dir_when_not_exists() {
        let dm = init_disk_dir_manager();

        let (_, test_dir, _t) = get_test_dir(dm.path_provider());

        let res = dm.get_dir(&test_dir);
        assert!(res.is_err());
    }

    #[test]
    fn test_disk_dir_manager_get_dir_when_exists() {
        let dm = init_disk_dir_manager();

        let (tmp_dir, test_dir, _t) = get_test_dir(dm.path_provider());
        dm.create_dir(&test_dir).unwrap();

        let res = dm.get_dir(&test_dir);
        assert!(res.is_ok());
        let dd = res.unwrap();
        assert_eq!("test_dir", dd.name);
        assert_eq!(tmp_dir, dd.path);
    }

    #[test]
    fn test_disk_dir_manager_is_dir_exists_when_not_exists() {
        let dm = init_disk_dir_manager();

        let (_, test_dir, _t) = get_test_dir(dm.path_provider());

        let res = dm.is_dir_exists(&test_dir);
        assert!(res.is_ok());
        assert!(!res.unwrap());
    }

    #[test]
    fn test_disk_dir_manager_is_dir_exists_when_exists() {
        let dm = init_disk_dir_manager();

        let (_, test_dir, _t) = get_test_dir(dm.path_provider());
        dm.create_dir(&test_dir).unwrap();

        let res = dm.is_dir_exists(&test_dir);
        assert!(res.is_ok());
        assert!(res.unwrap());
    }

    #[test]
    fn test_disk_dir_manager_delete_dir_when_not_exists() {
        let dm = init_disk_dir_manager();

        let (_, test_dir, _t) = get_test_dir(dm.path_provider());

        let res = dm.delete_dir(&test_dir, true);
        assert!(res.is_err());
    }

    #[test]
    fn test_disk_dir_manager_delete_dir_when_exists() {
        let dm = init_disk_dir_manager();

        let (_, test_dir, _t) = get_test_dir(dm.path_provider());
        dm.create_dir(&test_dir).unwrap();

        let res = dm.delete_dir(&test_dir, true);
        assert!(res.is_ok());

        let (_, test_dir, _t1) = get_test_dir(dm.path_provider());
        dm.create_dir(&test_dir).unwrap();

        let res = dm.delete_dir(&test_dir, false);
        assert!(res.is_ok());
    }
}

mod test_disk_file_manager_test {
    use super::*;

    #[test]
    fn test_disk_file_manager_path_provider() {
        let spp = SystemPathProvider::new();
        let dm = DiskDirectoryManager::new(spp);

        let fm = DiskFileManager::new(dm);

        let provider = fm.path_provider().as_any()
            .downcast_ref::<SystemPathProvider>()
            .unwrap();

        assert_eq!(&spp, provider);
    }

    #[test]
    fn test_disk_file_manager_dir_manager() {
        let dm = init_disk_dir_manager();
        let fm = DiskFileManager::new(dm);

        let dir_manager = fm.dir_manager().as_any()
            .downcast_ref::<DiskDirectoryManager<SystemPathProvider>>()
            .unwrap();

        assert_eq!(&dm, dir_manager);
    }

    #[tokio::test]
    async fn test_disk_file_manager_create_file_async_when_not_exists() {
        let fm = init_disk_file_manager();

        let (tmp_dir, test_file, _t) = get_test_file(fm.path_provider());

        let res = fm.create_file_async(&test_file).await;
        assert!(res.is_ok());
        let fd = res.unwrap();
        assert_eq!("test_file", fd.name);
        assert_eq!(tmp_dir, fd.path);
        assert_eq!(0, fd.len);
    }

    #[tokio::test]
    async fn test_disk_file_manager_create_file_async_when_exists() {
        let fm = init_disk_file_manager();

        let (_, test_file, _t) = get_test_file(fm.path_provider());
        fm.create_file_async(&test_file).await.unwrap();

        let res = fm.create_file_async(&test_file).await;
        assert!(res.is_err());
    }

    #[tokio::test]
    async fn test_disk_file_manager_get_file_async_when_not_exists() {
        let fm = init_disk_file_manager();

        let (_, test_file, _t) = get_test_file(fm.path_provider());

        let res = fm.get_file_async(&test_file).await;
        assert!(res.is_err());
    }

    #[tokio::test]
    async fn test_disk_file_manager_get_file_async_when_exists() {
        let fm = init_disk_file_manager();

        let (tmp_dir, test_file, _t) = get_test_file(fm.path_provider());
        fm.create_file_async(&test_file).await.unwrap();

        let res = fm.get_file_async(&test_file).await;
        assert!(res.is_ok());
        let fd = res.unwrap();
        assert_eq!("test_file", fd.name);
        assert_eq!(tmp_dir, fd.path);
        assert_eq!(0, fd.len);
    }

    #[tokio::test]
    async fn test_disk_file_manager_is_file_exists_async_when_not_exists() {
        let fm = init_disk_file_manager();

        let (_, test_file, _t) = get_test_file(fm.path_provider());

        let res = fm.is_file_exists_async(&test_file).await;
        assert!(res.is_ok());
        assert!(!res.unwrap());
    }

    #[tokio::test]
    async fn test_disk_file_manager_is_file_exists_async_when_exists() {
        let fm = init_disk_file_manager();

        let (_, test_file, _t) = get_test_file(fm.path_provider());
        fm.create_file_async(&test_file).await.unwrap();

        let res = fm.is_file_exists_async(&test_file).await;
        assert!(res.is_ok());
        assert!(res.unwrap());
    }

    #[tokio::test]
    async fn test_disk_file_manager_delete_file_async_when_not_exists() {
        let fm = init_disk_file_manager();

        let (_, test_file, _t) = get_test_file(fm.path_provider());

        let res = fm.delete_file_async(&test_file).await;
        assert!(res.is_err());
    }

    #[tokio::test]
    async fn test_disk_file_manager_delete_file_async_when_exists() {
        let fm = init_disk_file_manager();

        let (_, test_file, _t) = get_test_file(fm.path_provider());
        fm.create_file_async(&test_file).await.unwrap();

        let res = fm.delete_file_async(&test_file).await;
        assert!(res.is_ok());
    }
    
    #[test]
    fn test_disk_file_manager_create_file_when_not_exists() {
        let fm = init_disk_file_manager();

        let (tmp_dir, test_file, _t) = get_test_file(fm.path_provider());

        let res = fm.create_file(&test_file);
        assert!(res.is_ok());
        let fd = res.unwrap();
        assert_eq!("test_file", fd.name);
        assert_eq!(tmp_dir, fd.path);
        assert_eq!(0, fd.len);
    }

    #[test]
    fn test_disk_file_manager_create_file_when_exists() {
        let fm = init_disk_file_manager();

        let (_, test_file, _t) = get_test_file(fm.path_provider());
        fm.create_file(&test_file).unwrap();

        let res = fm.create_file(&test_file);
        assert!(res.is_err());
    }

    #[test]
    fn test_disk_file_manager_get_file_when_not_exists() {
        let fm = init_disk_file_manager();

        let (_, test_file, _t) = get_test_file(fm.path_provider());

        let res = fm.get_file(&test_file);
        assert!(res.is_err());
    }

    #[test]
    fn test_disk_file_manager_get_file_when_exists() {
        let fm = init_disk_file_manager();

        let (tmp_dir, test_file, _t) = get_test_file(fm.path_provider());
        fm.create_file(&test_file).unwrap();

        let res = fm.get_file(&test_file);
        assert!(res.is_ok());
        let fd = res.unwrap();
        assert_eq!("test_file", fd.name);
        assert_eq!(tmp_dir, fd.path);
        assert_eq!(0, fd.len);
    }

    #[test]
    fn test_disk_file_manager_is_file_exists_when_not_exists() {
        let fm = init_disk_file_manager();

        let (_, test_file, _t) = get_test_file(fm.path_provider());

        let res = fm.is_file_exists(&test_file);
        assert!(res.is_ok());
        assert!(!res.unwrap());
    }

    #[test]
    fn test_disk_file_manager_is_file_exists_when_exists() {
        let fm = init_disk_file_manager();

        let (_, test_file, _t) = get_test_file(fm.path_provider());
        fm.create_file(&test_file).unwrap();

        let res = fm.is_file_exists(&test_file);
        assert!(res.is_ok());
        assert!(res.unwrap());
    }

    #[test]
    fn test_disk_file_manager_delete_file_when_not_exists() {
        let fm = init_disk_file_manager();

        let (_, test_file, _t) = get_test_file(fm.path_provider());

        let res = fm.delete_file(&test_file);
        assert!(res.is_err());
    }

    #[test]
    fn test_disk_file_manager_delete_file_when_exists() {
        let fm = init_disk_file_manager();

        let (_, test_file, _t) = get_test_file(fm.path_provider());
        fm.create_file(&test_file).unwrap();

        let res = fm.delete_file(&test_file);
        assert!(res.is_ok());
    }
}

fn get_test_dir(provider: &dyn PathProvider) -> (String, String, TempDir) {
    let t = tempdir().unwrap();
    let tmp_dir = t.path().to_str().unwrap();
    let test_dir = provider.combine(vec![tmp_dir, "test_dir"]).unwrap();

    (tmp_dir.to_string(), test_dir, t)
}

fn get_test_file(provider: &dyn PathProvider) -> (String, String, TempDir) {
    let t = tempdir().unwrap();
    let tmp_dir = t.path().to_str().unwrap();
    let test_file = provider.combine(vec![tmp_dir, "test_file"]).unwrap();

    (tmp_dir.to_string(), test_file, t)
}

fn init_disk_dir_manager() -> DiskDirectoryManager<SystemPathProvider> {
    DiskDirectoryManager::new(SystemPathProvider::new())
}

fn init_disk_file_manager() -> DiskFileManager<DiskDirectoryManager<SystemPathProvider>> {
    DiskFileManager::new(init_disk_dir_manager())
}